Make Nd tensors hit fused addmm pass

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107003
* #107004
* #106912
* __->__ #106911

Replace https://github.com/pytorch/pytorch/pull/106433 since I had a bad cla commit.

Speeds up eager convnext bfloat16 inference by 35%., and eager timm bfloat16 inference average by `.5%`

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

eellison committed:
Make Nd tensors hit fused addmm pass

[ghstack-poisoned]

eellison commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -324,9 +324,10 @@ def forward(self, x, y):
+                mod_c = torch.compile(mod)
+                out, code = run_and_get_code(mod_c, v, other)
+                self.assertEqual(out, mod(v, other), rtol=1e-2, atol=1e-2)
+                # TODO - assert fusions work code
```
cc @jgong5, would you mind taking a look at this ?

eellison committed:
Update on "Make Nd tensors hit fused addmm pass"


Replace https://github.com/pytorch/pytorch/pull/106433 since I had a bad cla commit.

Speeds up eager convnext bfloat16 inference by 35%., and eager timm bfloat16 inference average by `.5%`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

ezyang commented on test/inductor/test_aot_inductor.py:
```diff
@@ -152,7 +152,6 @@ def __init__(self):
                 )
                 self.bn1 = torch.nn.BatchNorm2d(num_features=16)
                 self.relu1 = torch.nn.ReLU()
-                self.fc1 = torch.nn.Linear(in_features=1638400, out_features=1)
```
tell me more?

ezyang commented on aten/src/ATen/native/Linear.cpp:
```diff
@@ -6,8 +6,10 @@
+#include <c10/core/SymInt.h>
 #include <c10/util/MaybeOwned.h>
 #include <ATen/TensorSubclassLikeUtils.h>
+#include <iostream>
```
dead now right

ezyang commented on aten/src/ATen/native/Linear.cpp:
```diff
@@ -53,20 +55,29 @@ static inline bool parseLinearFlatten3d() {
-    return result.view_symint({input_sizes[0], input_sizes[1], result.sym_size(1)});
+    // can't use -1 in reshape because it errors when a dimension is 0
+    c10::SymInt flattened_dim = 1;
+    for (size_t i = 0, ndim = input_sizes.size(); i < ndim - 1; ++i) {
```
`ndim - 1` here hazardous because you are using unsigned `size_t` type, you will overflow if `ndim == 0`. Just use `int64_t`, the unsigned type here really is not worth it.

ezyang commented on aten/src/ATen/native/Linear.cpp:
```diff
@@ -53,20 +55,29 @@ static inline bool parseLinearFlatten3d() {
+    auto inp_reshape = input.reshape_symint({flattened_dim, input_sizes.at(input_sizes.size() -1)});
+    const auto result = at::addmm(bias, inp_reshape, weight.t());
+    auto new_size = input_sizes.slice(0, input_sizes.size() - 1);
+    std::vector<SymInt> sizes_vec(new_size.begin(), new_size.end());
```
Consider using SymDimVector to avoid the heap allocation

ezyang reviewed (approved):
okey dokey

eellison committed:
Update on "Make Nd tensors hit fused addmm pass"


Replace https://github.com/pytorch/pytorch/pull/106433 since I had a bad cla commit.

Speeds up eager convnext bfloat16 inference by 35%., and eager timm bfloat16 inference average by `.5%`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

eellison committed:
Update on "Make Nd tensors hit fused addmm pass"


Replace https://github.com/pytorch/pytorch/pull/106433 since I had a bad cla commit.

Speeds up eager convnext bfloat16 inference by 35%., and eager timm bfloat16 inference average by `.5%`

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov anijain2305

[ghstack-poisoned]

eellison commented:
@pytorchbot merge -f "unrelated failure"
